### PR TITLE
Add ModelicaDuplicateString and ModelicaDuplicateStringWithErrorReturn (new in Modelica Language Specification version 3.5)

### PR DESCRIPTION
**Pull request by beutlich**
Can only be merged once MSL is based on MLS 3.5.

### CODE REVIEW

**Review by henrikt-ma — COMMENTED** [review]
Looks good to me, assuming you looked at all occurrences of `ModelicaAllocateString` and `ModelicaAllocateStringWithErrorReturn` to find all that should be replaced by the new functions.

By the way, I noticed that the test implementations of the old functions still don't add a null terminator.  Is there a plan to also update these (and remove any redundant null termination on the results of these calls)?

**Comment by beutlich** [comment]
> By the way, I noticed that the test implementations of the old functions still don't add a null terminator. Is there a plan to also update these (and remove any redundant null termination on the results of these calls)?

You mean ModelicaAllocateString beahves like calloc? I missed that change.

**Comment by henrikt-ma** [comment]
> You mean ModelicaAllocateString beahves like calloc? I missed that change.

Not exactly.  It allows for detection of uninitialized reads by only zeroing the last byte.

**Review by dietmarw — CHANGES_REQUESTED** [review]
Blocking accidental merge. Need to wait until master is based on MLS 3.~~5~~6.

**Comment by HansOlsson** [comment]
> Blocking accidental merge. Need to wait until master is based on MLS 3.5.

This has now been changed (to 3.6 to be specific).
@dietmarw can you check again?

**Comment by beutlich** [comment]
> > Blocking accidental merge. Need to wait until master is based on MLS 3.5.
> 
> This has now been changed (to 3.6 to be specific). @dietmarw can you check again?

So, others features than 3.6 annotations are desired, right?

**Comment by henrikt-ma** [comment]
> So, others features than 3.6 annotations are desired, right?

Yes.  Tool vendors will need to react and bring up a discussion if they think the adoption of a particular new feature (such as these string allocation functions) is too fast.

**Comment by dietmarw** [comment]
It would help if there was a formal decision what MSL version would now allow MLS 3.6 features so the milestone of this PR can be set accordingly.

**Comment by maltelenz** [comment]
> It would help if there was a formal decision what MSL version would now allow MLS 3.6 features so the milestone of this PR can be set accordingly.

There was (at the monthly MAP-Lib meeting on the 14th of November, if I remember the date right). 4.1.0 will allow 3.6 features. As Henrik says, if any tool vendor thinks a specific feature from 3.6 is not ready to be used, they can speak up in the corresponding pull request.

**Review by HansOlsson — APPROVED** [review]
Looks good.

File: .CI/Test/Common.c [hunk 1]
```diff
@@ -13,6 +14,23 @@ char* ModelicaAllocateStringWithErrorReturn(size_t len) {
     return malloc(len + 1); /* Never free'd in the test programs */
 }
 
+char* ModelicaDuplicateString(const char* str) {
+    void *data = malloc(strlen(str) + 1); /* Never free'd in the test programs */
+    assert(data);
+    if (NULL != data) {
+        strcpy(data, str);
+    }
```
Comment (henrikt-ma) [inline .CI/Test/Common.c:22]: As far as I can tell, we're already relying on `assert` not being disabled, so the difference to `ModelicaDuplicateStringWithErrorReturn` could be made more clear by removing the redundant guard:
```suggestion
      strcpy(data, str);
```

File: .CI/Test/ModelicaUtilities.h [hunk 2]
```diff
@@ -201,6 +195,24 @@ resources use ModelicaError or ModelicaFormatError to signal
...
+provided by the calling program, as for any other array. If an error
+occurs, this function does not return, but calls "ModelicaError".
+*/
+
+char* ModelicaDuplicateStringWithErrorReturn(const char* str);
+/*
+Same as ModelicaDuplicateString, except that in case of error, the
+function returns 0. This allows the external function to close files
+and free other open resources in case of error. After cleaning up
+resources use ModelicaError or ModelicaFormatError to signal
```
Comment (henrikt-ma) [inline .CI/Test/ModelicaUtilities.h:212]: Comma for readability:
```suggestion
resources use, ModelicaError or ModelicaFormatError to signal
```

File: .CI/Test/Common.c [hunk 3]
```diff
@@ -13,6 +14,23 @@ char* ModelicaAllocateStringWithErrorReturn(size_t len) {
     return malloc(len + 1); /* Never free'd in the test programs */
 }
 
+char* ModelicaDuplicateString(const char* str) {
+    void *data = malloc(strlen(str) + 1); /* Never free'd in the test programs */
+    assert(data);
```
Comment (henrikt-ma) [inline .CI/Test/Common.c:19]: The documentation comment in the header promises to call `ModelicaError` in case of failure.

Comment (henrikt-ma) [reply .CI/Test/Common.c:19]: Maybe the best solution would be to use `ModelicaAllocateString` for allocation, and then make sure `ModelicaAllocateString` actually calls `ModelicaError`?

Comment (beutlich) [reply .CI/Test/Common.c:19]: Well, already wrong in ModelicaAllocateString.

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
Not that it matters a whole lot, but I think it would be possible to call `ModelicaError` (which ends with `assert(0)` anyways) as promised in case of error.

**Review by henrikt-ma — APPROVED** [review]
Looks good to me.

File: .CI/Test/Common.c (hunk 1 above)
Comment (beutlich) [reply .CI/Test/Common.c]: Sure, it is all about test code only.

**Pull request merged**